urlc945a4be7c_dburggie_jatrace_tgz-pJ8-renders_EllipseJ8/src/jatrace/LinkedBody.java

```diff
@@ -0,0 +1,89 @@
+package jatrace;
+
+/** This class provides a means to keep track of a collection of objects with
+ *  the Body interface through a linked list. Each link will have a unique ID
+ *  integer that can be used to find and or remove it from the collection. */
+public class LinkedBody
+{
+
+	private LinkedBody nextLink;
+	private LinkedBody prevLink;
+	private Body body;
+	
+	
+	/** Instantiates a new link containing the specified Body interface. */
+	public LinkedBody(Body b)
+	{
+		body = b;
+		nextLink = null;
+		prevLink = null;
+	}
+	
+	/** Fetches a reference to the Body object this link wraps. */
+	public Body b()
+	{
+		return body;
+	}
+	
+	/** Fetches the next link down from caller. A null reference will be
+	 *  returned if the caller is the tail object. */
+	public LinkedBody next()
+	{
+		return nextLink;
+	}
+	
+	/** Removes object from its linked list and inserts it before the argument.
+	 *  This function safely handles null references. */
+	public void insertBefore(LinkedBody lb)
+	{
+		//clear this objects ends
+		remove();
+		
+		if (lb != null)
+		{
+			prevLink = lb.prevLink;
+			lb.prevLink = this;
+			nextLink = lb;
+			if (prevLink != null)
+			{
+				prevLink.nextLink = this;
+			}
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 40:51, 58:69.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.